src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.JSInterop.Infrastructure
+{
+    internal sealed class ByteArrayJsonConverter : JsonConverter<byte[]>
+    {
+        // Unique identifier for serializing byte arrays.
+        private int _byteArrayId;
+
+        internal static readonly JsonEncodedText ByteArrayRefKey = JsonEncodedText.Encode("__byte[]");
+
+        public ByteArrayJsonConverter(JSRuntime jSRuntime)
+        {
+            JSRuntime = jSRuntime;
+        }
+
+        public JSRuntime JSRuntime { get; }
+
+        public override byte[]? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (JSRuntime.ByteArraysToBeRevived.Count == 0)
+            {
+                throw new JsonException("ByteArraysToBeRevived is empty.");
+            }
+
+            int? byteArrayRef = null;
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType == JsonTokenType.PropertyName)
+                {
+                    if (byteArrayRef is null && reader.ValueTextEquals(ByteArrayRefKey.EncodedUtf8Bytes))
+                    {
+                        reader.Read();
+                        byteArrayRef = reader.GetInt32();
+                    }
+                    else
+                    {
+                        throw new JsonException($"Unexpected JSON property {reader.GetString()}.");
+                    }
+                }
+                else
+                {
+                    throw new JsonException($"Unexpected JSON Token {reader.TokenType}.");
+                }
+            }
```

Comment:
Why were we unable to use the `Serialize` trick here?

If we are parsing this in this way, can we be more strict and check for the `exact` sequence here? I believe we should be able to check something like
1. Start object
2. __byte[] property
3. End object

And for anything else we can/should fail. 

Also, if we are parsing the input like this, we need to write tests for how it behaves against malformed inputs like `{ __byte[]: "something" }`, etc.

Reply:
> Why were we unable to use the Serialize trick here?

The `Serialize` trick causes the WASM E2E tests to fail in the CI. Not able to repro the issue locally, and tried logging/several different solutions without any luck. This isn't really efficient given each attempt needs to be run through the entire CI pipeline to see if it works.

> If we are parsing this in this way, can we be more strict and check for the exact sequence here?
> Also, if we are parsing the input like this, we need to write tests for how it behaves against malformed inputs like { __byte[]: "something" }, etc.

Sure thing, updated. 02b2b75ec1